### PR TITLE
Add AMP.pushState() action

### PR DESCRIPTION
**Pull request by choumx**
Partial for #9128.

- Add `AMP.pushState()` action which allows "storing" amp-bind variables in history stack.
- Add max history stack size of 10. Over-cap callbacks are pruned on pop plus `user().warn`.

TODO:
- Support browser restarts/restoring tabs.
- Support forward navigation.

### CODE REVIEW

File: extensions/amp-bind/0.1/bind-impl.js [hunk 1]
```diff
@@ -214,25 +218,38 @@ export class Bind {
...
+        .then(result => this.setState(result));
     return this.setStatePromise_;
   }
 
   /**
+   * Same as setStateWithExpression() except also pushes new history.
+   * Popping the new history stack entry will restore the values of variables
+   * in `expression`.
+   * @param {string} expression
+   * @param {!JsonObject} scope
```
Comment (aghassemi) [inline extensions/amp-bind/0.1/bind-impl.js]: scope type is inconsistent among setStateWithExpression, pushStateWithExpression and evaluateExpression_

File: extensions/amp-bind/0.1/bind-impl.js [hunk 2]
```diff
@@ -214,25 +218,38 @@ export class Bind {
...
+   * @return {!Promise}
+   */
+  pushStateWithExpression(expression, scope) {
+    return this.evaluateExpression_(expression, scope).then(result => {
+      // Store the current values of each referenced variable in `expression`
+      // so that we can restore them on history-pop.
+      const oldState = map();
+      Object.keys(result).forEach(variable => {
+        const s = map();
+        s[variable] = this.state_[variable] || null; // `undefined` -> `null`.
```
Comment (aghassemi) [inline extensions/amp-bind/0.1/bind-impl.js]: are values in `state_` always object? (making sure we are not turning `0`, `false`, `''` to `null`)

File: extensions/amp-bind/0.1/test/integration/test-bind-impl.js [hunk 3]
```diff
@@ -104,7 +109,8 @@ describe.configure().ifChrome().skipOldChrome().run('Bind', function() {
   const TIMEOUT = Math.max(window.ampTestRuntimeConfig.mochaTimeout, 4000);
   this.timeout(TIMEOUT);
 
-  describe.configure().skipSauceLabs().run('in FIE', function() {
+  // TODO(choumx): Fix and unskip.
+  describe.configure().skipSauceLabs().run('(skip)', function() {
```
Comment (aghassemi) [inline extensions/amp-bind/0.1/test/integration/test-bind-impl.js]: Does `skipSauceLabs()` work anymore? I've seen be be replaced by `ifChrome().skipOldChrome()``

File: extensions/amp-bind/0.1/amp-state.js [hunk 4]
```diff
@@ -146,7 +147,7 @@ export class AmpState extends AMP.BaseElement {
       return;
     }
     const id = user().assert(this.element.id, '<amp-state> must have an id.');
-    const state = Object.create(null);
+    const state = /** @type {!JsonObject} */ (map());
```
Comment (choumx) [inline extensions/amp-bind/0.1/amp-state.js]: The type is correct but returns `{}`, not a proto-less object. I'll see if I can change that in `dict()` itself in a follow-up.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 1 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Fixed, thanks.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 2 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Woops, good catch!

File: extensions/amp-bind/0.1/test/integration/test-bind-impl.js (hunk 3 above)
Comment (choumx) [inline extensions/amp-bind/0.1/test/integration/test-bind-impl.js]: Raghu's going to address this.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 5]
```diff
@@ -210,29 +214,43 @@ export class Bind {
...
+      const oldState = map();
+      Object.keys(result).forEach(variable => {
+        const s = map();
+        const value = this.state_[variable];
+        s[variable] = (value === undefined) ? null : value;
+        deepMerge(oldState, s, MAX_MERGE_DEPTH);
+      });
+
+      const onPop = () => this.setState(oldState);
+      this.history_.push(onPop);
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: @choumx As I mentioned, my main concern is reducing number of pushes from "indefinite". Do you have a plan to limit these?

Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Added `maxStackSize_()` of 10 which ignores and prunes over-cap callbacks on pop. We could also do this during push, but this avoids an extra iteration over the callbacks array.

Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: Wait, but why is it on pop, and not push?

Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Per above, it avoid iterating over the callbacks array on every push. It has the same functional outcome. 

I assumed this was about UX. Are you concerned about memory usage?

Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: Not memory. Concerned with this scenario: user clicks in SRP, taps N times and then it's too hard to find the way back. Normally back button closes the doc and returns back to SRP after a tap or two. Making it N taps will still be ok, as long as there's some limit. So, from what I see, i'd expect we'd limit on push and not pop. No?

File: extensions/amp-bind/0.1/bind-impl.js [hunk 6]
```diff
@@ -210,29 +214,43 @@ export class Bind {
...
+  pushStateWithExpression(expression, scope) {
+    return this.evaluateExpression_(expression, scope).then(result => {
+      // Store the current values of each referenced variable in `expression`
+      // so that we can restore them on history-pop.
+      const oldState = map();
+      Object.keys(result).forEach(variable => {
+        const s = map();
+        const value = this.state_[variable];
+        s[variable] = (value === undefined) ? null : value;
+        deepMerge(oldState, s, MAX_MERGE_DEPTH);
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: Why is this deep merging? Couldn't this just be simple assignment to `oldState[s] = value`?

File: extensions/amp-bind/0.1/bind-impl.js (hunk 5 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Oh I see. Don't users have the large "X" button in the top left to return to SRP? I'm not sure we should install a general constraint for a single AMP surface (SRP), especially as we push AMP embeds, AMP as canonical, etc.

Implementation would also need to be complicated to support browser restarts and restoring tabs (storing current stack size in memory won't suffice). 

I'll start a separate discussion issue for this and loop in the PMs.

**Comment by choumx** [comment]
One thing I neglected is the case of browser restarts/restoring tabs. `amp-bind` also needs to solve this regardless of history manipulation.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 5 above)
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: I'm ok if we rollback this for now and continue discussion on #11071

Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Done, thanks. I won't publish documentation for this feature until we resolve that discussion.

Any other comments for this PR?

Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: I'm all set.

**Pull request merged**